### PR TITLE
test: Add tests for a LSP6 controlled Mintable LSP7

### PR DESCRIPTION
**Pull request by b00ste**
## What does this PR introduce?

In this pr we create a new test file that tests the behavior of a LSP6 controlled Mintable LSP7.

Methods:

- using execute(..) in LSP7 through LSP6 --> reverts
- using mint(..) in LSP7 through LSP6 --> reverts
- using renounceOwnership(..) in LSP7 through LSP6 --> reverts
- using transferOwnership(..) in LSP7 through LSP6 --> works
- using setData(..) in lSP7 through LSP6 --> works

Permissions:

- CHANGEOWNER --> works
- CHANGEPERMISSIONS --> works 
- ADDPERMISSIONS --> works
- SETDATA --> works
- CALL --> not working
- STATICCALL --> not working
- DELEGATECALL --> not working
- DEPLOY --> not working
- TRANSFERVALUE --> not working
- SIGN --> works
- ENCRYPT --> not tested

### CODE REVIEW

File: tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts [hunk 1]
```diff
@@ -0,0 +1,840 @@
...
+  );
+
+  const keyManager = await new LSP6KeyManager__factory(accounts[0]).deploy(
+    lsp7.address
+  );
+
+  const keys = [
+    ERC725YKeys.LSP6["AddressPermissions[]"].length,
+    ERC725YKeys.LSP6["AddressPermissions[]"].index + ARRAY_INDEX.ZERO,
+    ERC725YKeys.LSP6["AddressPermissions:Permissions"] +
```
Comment (CJ42) [inline tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts:48]: maybe you can store the `address[0]` in a `owner` variable (for the address that has ALL PERMISSIONS)?

File: tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts [hunk 2]
```diff
@@ -0,0 +1,840 @@
...
+
+describe("When deploying LSP7 with LSP6 as owner", () => {
+  let context: LSP6ControlledToken;
+
+  before(async () => {
+    context = await buildContext();
+  });
+
+  it("should have lsp6 as owner of the lsp7", async () => {
+    expect(await context.token.owner()).to.be.equal(context.keyManager.address);
```
Comment (CJ42) [inline tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts:96]:
```suggestion
    expect(await context.token.owner()).to.equal(context.keyManager.address);
```

File: tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts [hunk 3]
```diff
@@ -0,0 +1,840 @@
...
+        context.accounts[0].address.substring(2),
+    ];
+
+    const values = [
+      ARRAY_LENGTH.ONE,
+      context.accounts[0].address,
+      ALL_PERMISSIONS,
+    ];
+
+    expect(await context.token["getData(bytes32[])"](keys)).to.be.deep.equal(
```
Comment (CJ42) [inline tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts:113]:
```suggestion
    expect(await context.token["getData(bytes32[])"](keys)).to.deep.equal(
```

File: tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts [hunk 4]
```diff
@@ -0,0 +1,840 @@
...
+      values
+    );
+  });
+
+  describe("using execute(..) in LSP7 through LSP6", () => {
```
Comment (CJ42) [inline tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts:118]:
```suggestion
  describe("when trying to call execute(..) function on LSP7 through LSP6", () => {
```

File: tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts [hunk 5]
```diff
@@ -0,0 +1,840 @@
...
+    it("should revert", async () => {
```
Comment (CJ42) [inline tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts:119]: Maybe we could add a bit more descriptions in the tests, so that we know why it should revert?

```suggestion
    it("should revert because function does not exist on LSP7", async () => {
```

File: tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts [hunk 6]
```diff
@@ -0,0 +1,840 @@
...
+      const newTokenContract = await new LSP7Tester__factory(
+        context.accounts[0]
+      ).deploy("NewTokenName", "NewTokenSymbol", context.accounts[0].address);
+      const mintPayload = newTokenContract.interface.encodeFunctionData(
```
Comment (CJ42) [inline tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts:123]: How about adding a comment above, to explain that we try to call the `mint(...)` function in another contract? (as we were discussing)

File: tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts [hunk 7]
```diff
@@ -0,0 +1,840 @@
...
+        .connect(context.accounts[0])
+        .execute(payload);
+
+      await expect(executePayload).to.be.revertedWith(
+        "LSP6: Unknown Error occured when calling the linked target contract"
+      );
+    });
+  });
+
+  describe("using mint(..) in LSP7 through LSP6", () => {
```
Comment (CJ42) [inline tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts:144]: same comment as above.

File: tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts [hunk 8]
```diff
@@ -0,0 +1,840 @@
...
+        1,
+        true,
+        "0x",
+      ]);
+
+      const executeMintPayload = context.keyManager
+        .connect(context.accounts[0])
+        .execute(mintPayload);
+
+      await expect(executeMintPayload)
```
Comment (CJ42) [inline tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts:158]: I would suggest to put the whole call inside the `expect(...)`, as the tests sometimes are buggy when testing events like that.

File: tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts [hunk 9]
```diff
@@ -0,0 +1,840 @@
...
+        .to.be.revertedWithCustomError(
+          context.keyManager,
+          "InvalidERC725Function"
+        )
+        .withArgs(renounceOwnershipPayload);
+    });
+  });
+
+  describe("using transferOwnership(..) in LSP7 through LSP6", () => {
+    it("should change the owner of LSP7 contract", async () => {
```
Comment (CJ42) [inline tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts:186]: We could add some extra tests here, to ensure that afterwards, to ensure that:

- the functions marked as `onlyOwner` on the LSP7 token contract cannot be called by the KeyManager anymore afterwards.
- only the new owner (the EOA) can call these functions.

File: tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts [hunk 10]
```diff
@@ -0,0 +1,840 @@
...
+      const transferOwnershipPayload =
+        context.token.interface.encodeFunctionData("transferOwnership", [
+          context.accounts[0].address,
```
Comment (CJ42) [inline tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts:189]: It is a bit misleading here to use the same address as the EOA that had all the permissions previously (`accounts[0]`).

Maybe we could use an other account to make it more explicit? Like:

```js
const newOwner = context.accounts[1];

const transferOwnershipPayload =
        context.token.interface.encodeFunctionData("transferOwnership", [
          newOwner.address,
        ]);

```

File: tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts [hunk 11]
```diff
@@ -0,0 +1,840 @@
...
+        ]);
+
+      await context.keyManager
+        .connect(context.accounts[0])
+        .execute(transferOwnershipPayload);
+
+      expect(await context.token.owner()).to.equal(context.accounts[0].address);
+    });
+
+    it("should revert when calling setData(..)", async () => {
```
Comment (CJ42) [inline tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts:199]: ah ok this is done here. Maybe we can improve the description.

Or use a `describe` block?

```js
describe("when previous owner try to interact via the Key Manager", () => {
    it("`setData(...)` -> should revert with 'caller is not the owner' error.", async () => {
        // ....
    }) 
})
```

File: tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts [hunk 12]
```diff
@@ -0,0 +1,840 @@
...
+          ARRAY_LENGTH.TWO,
+          context.accounts[0].address,
+          context.accounts[1].address,
+          ALL_PERMISSIONS,
+          PERMISSIONS.CHANGEOWNER,
+        ];
+
+        expect(
+          await context.token["getData(bytes32[])"](keys)
+        ).to.be.deep.equal(values);
```
Comment (CJ42) [inline tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts:310]:
```suggestion
        ).to.deep.equal(values);
```

File: tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts [hunk 13]
```diff
@@ -0,0 +1,840 @@
...
+        await context.keyManager
+          .connect(context.accounts[1])
+          .execute(transferOwnershipPayload);
+
+        expect(await context.token.owner()).to.equal(
+          context.accounts[1].address
+        );
+      });
+
+      after(async () => {
```
Comment (CJ42) [inline tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts:359]: very nice 👌

File: tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts [hunk 14]
```diff
@@ -0,0 +1,840 @@
...
+          ALL_PERMISSIONS,
+          PERMISSIONS.CHANGEOWNER,
+        ];
+
+        expect(
+          await context.token["getData(bytes32[])"](keys)
+        ).to.be.deep.equal(values);
+      });
+
+      it("should revert when trying to use `renounceOwnership()`", async () => {
```
Comment (CJ42) [inline tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts:313]: maybe we should not put this test here, as it is not related to the permission CHANGEOWNER.
Because renounceOwnership cannot be called via the KeyManager (the KM does not provide this path).

So maybe in a separate describe block for edge cases?

File: tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts [hunk 15]
```diff
@@ -0,0 +1,840 @@
...
+          context.accounts[1].address,
+          context.accounts[2].address,
+          ALL_PERMISSIONS,
+          PERMISSIONS.CHANGEOWNER,
+          PERMISSIONS.CHANGEPERMISSIONS,
+        ];
+
+        expect(
+          await context.token["getData(bytes32[])"](keys)
+        ).to.be.deep.equal(values);
```
Comment (CJ42) [inline tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts:403]:
```suggestion
        ).to.deep.equal(values);
```

File: tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts [hunk 16]
```diff
@@ -0,0 +1,840 @@
...
+          ERC725YKeys.LSP6["AddressPermissions:Permissions"] +
+          context.accounts[0].address.substring(2);
+        const value = PERMISSIONS.CALL;
+        const payload = context.token.interface.encodeFunctionData(
+          "setData(bytes32,bytes)",
+          [key, value]
+        );
+
+        const changePermission = context.keyManager
+          .connect(context.accounts[1])
```
Comment (CJ42) [inline tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts:417]: maybe we can define this signer as `addressCanChangePermission`?
Otherwise, it is difficult to infer which permissions the caller has by reading the test.

File: tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts [hunk 17]
```diff
@@ -0,0 +1,840 @@
...
+          "setData(bytes32,bytes)",
+          [key, value]
+        );
+
+        await context.keyManager.connect(context.accounts[2]).execute(payload);
+
+        expect(await context.token["getData(bytes32)"](key)).to.equal(value);
+      });
+
+      it("should add back ALL_PERMISSIONS of the address", async () => {
```
Comment (CJ42) [inline tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts:440]: This is actually the wrong behaviour that we will have to fix.
It should be possible if you have CHANGEPERMISSIONS permission to add more permissions that it is initially.

See this comment: https://github.com/lukso-network/LIPs/pull/102#issuecomment-1205535756

As a note. Will have to update in the future.

File: tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts [hunk 18]
```diff
@@ -0,0 +1,840 @@
...
+          PERMISSIONS.CHANGEPERMISSIONS,
+          PERMISSIONS.ADDPERMISSIONS,
+        ];
+
+        expect(
+          await context.token["getData(bytes32[])"](keys)
+        ).to.be.deep.equal(values);
+      });
+
+      it("should be allowed to add permissions", async () => {
```
Comment (CJ42) [inline tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts:501]: I suggest this wording, to make it more clear.

```suggestion
      it("should be allowed to add a new controller address with some permissions", async () => {
```

File: tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts [hunk 19]
```diff
@@ -0,0 +1,840 @@
...
+          "setData(bytes32,bytes)",
+          [key, value]
+        );
+
+        await context.keyManager.connect(context.accounts[3]).execute(payload);
+
+        expect(await context.token["getData(bytes32)"](key)).to.equal(value);
+      });
+
+      it("should revert when trying to remove permissions", async () => {
```
Comment (CJ42) [inline tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts:516]: same here.

```suggestion
      it("should not be authorized to edit permissions of an existing controller (e.g: removing permissions)", async () => {
```

File: tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts [hunk 20]
```diff
@@ -0,0 +1,840 @@
...
+        const payload = context.token.interface.encodeFunctionData(
+          "setData(bytes32,bytes)",
+          [key, value]
+        );
+
+        const executePayload = context.keyManager
+          .connect(context.accounts[3])
+          .execute(payload);
+
+        await expect(executePayload)
```
Comment (CJ42) [inline tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts:530]: Maybe we can put the contract call direct inside the `expect` here?
Since we don't use the `const` defined.

File: tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts [hunk 21]
```diff
@@ -0,0 +1,840 @@
...
+          ALL_PERMISSIONS,
+          PERMISSIONS.CHANGEOWNER,
+          PERMISSIONS.CHANGEPERMISSIONS,
+          PERMISSIONS.ADDPERMISSIONS,
+          PERMISSIONS.SETDATA,
+        ];
+
+        expect(
+          await context.token["getData(bytes32[])"](keys)
+        ).to.be.deep.equal(values);
```
Comment (CJ42) [inline tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts:596]: Can you check that the other tests don't have the `be` like here, but just `to.deep.equal(...)`

```suggestion
        ).to.deep.equal(values);
```

File: tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts [hunk 22]
```diff
@@ -0,0 +1,840 @@
...
+      });
+
+      it("should allow second controller to use setData", async () => {
```
Comment (CJ42) [inline tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts:599]: nice test 👌

File: tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts [hunk 23]
```diff
@@ -0,0 +1,840 @@
...
+      it("should restrict second controller with AllowedERC725YKeys", async () => {
+        const key =
+          ERC725YKeys.LSP6["AddressPermissions:AllowedERC725YKeys"] +
+          context.accounts[4].address.substring(2);
+        const value = abiCoder.encode(
+          ["bytes32[]"],
+          [
+            [
+              ethers.utils
+                .keccak256(ethers.utils.toUtf8Bytes("FirstRandomString"))
```
Comment (CJ42) [inline tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts:625]: You can define this variable in a constant at the beginning of the `describe` block and reuse it across tests (define the data key prefix in a constant for instance).

File: tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts [hunk 24]
```diff
@@ -0,0 +1,840 @@
...
+        await context.keyManager.connect(context.accounts[4]).execute(payload);
+
+        expect(
+          await context.token["getData(bytes32[])"](keys)
+        ).to.be.deep.equal(values);
+      });
+
+      it("should revert when trying to change a key that is not allowed", async () => {
+        const keys = [
+          ethers.utils.keccak256(ethers.utils.toUtf8Bytes("FirstRandomString")),
```
Comment (CJ42) [inline tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts:685]: not sure to see here which data key is not allowed.

I guess the first one because it does not have 16 x `00`s? Maybe you could hash another word to make this more explicit.

```suggestion
          ethers.utils.keccak256(ethers.utils.toUtf8Bytes("Not Allowed ERC725Y data key")),
```

File: tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts [hunk 25]
```diff
@@ -0,0 +1,840 @@
...
+    /**
+     * Testing the following permissions is skipped because
+     * execute(..) is not available in a token contract.
+     *
+     * CALL
+     * STATICCALL
+     * DELEGATECALL
+     * DEPLOY
+     * TRANSFERVALUE
+     */
```
Comment (CJ42) [inline tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts:756]: How about moving the describe block when testing `execute(...)` here?
We could also add the other tests for CALL, DELEGATECALL, STATICCALL, DEPLOY and TRANSFERVALUE in the future. But this is overkill for now.

**Review by CJ42 — CHANGES_REQUESTED** [review]
Nice testing for some potential future use cases! I have added some review comments.

File: tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts (hunk 14 above)
Comment (b00ste) [reply tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts:313]: you’re right and the test already exists on line 170

**Review by YamenMerhi — APPROVED** [review]

**Review by CJ42 — APPROVED** [review]
